File: snackbar/src/main/java/com/snackbar/SnackBuilder.java
package com.snackbar;

import android.app.Activity;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;

/**
 * Created by clickapps on 2/11/17.
 */

public class SnackBuilder {

    private SnackBarParam param;
    private Snackbar snackbar;

    public SnackBuilder(@NonNull Activity context, @StringRes int resId) {
        param = new SnackBarParam();
        param.context = context;
        param.message = context.getString(resId);
    }

    public SnackBuilder(@NonNull Activity context, @NonNull String msg) {
        param = new SnackBarParam();
        param.context = context;
        param.message = msg;
    }

    public SnackBuilder(@NonNull Activity context, @StringRes int resId, @ColorRes int backgroundColor) {
        param = new SnackBarParam();
        param.context = context;
        param.message = context.getString(resId);
        param.actionBackgroundResId = backgroundColor;
    }

    public SnackBuilder(@NonNull Activity context, String msg, @ColorRes int backgroundColor) {
        param = new SnackBarParam();
        param.context = context;
        param.message = msg;
        param.actionBackgroundResId = backgroundColor;
    }

    public SnackBuilder actionMessage(String actionMessage) {
        param.actionMessage = actionMessage;
        return this;
    }

    public SnackBuilder actionMessage(@StringRes int resId) {
        param.actionMessage = param.context.getString(resId);
        return this;
    }

    public SnackBuilder view(View view) {
        param.snackBarView = view;
        param.snackBarView.setFocusable(false);
        return this;
    }

    public SnackBuilder listener(OnSnackBarActionListener l) {
        param.onSnackBarActionListener = l;
        return this;
    }

    public SnackBuilder actionColor(@ColorRes int resId) {
        param.actionColorResId = resId;
        return this;
    }

    public SnackBuilder textColor(@ColorRes int resId) {
        param.textColor = resId;
        return this;
    }

    public SnackBuilder duration(int duration) {
        param.snackBarDuration = duration;
        return this;
    }

    public SnackBuilder backgroundColor(@ColorRes int color) {
        param.actionBackgroundResId = color;
        return this;
    }

    public SnackBuilder info() {
        param.actionBackgroundResId = SnackBarParam.blue;
        return this;
    }

    public SnackBuilder confirm() {
        param.actionBackgroundResId = SnackBarParam.green;
        return this;
    }

    public SnackBuilder warning() {
        param.actionBackgroundResId = SnackBarParam.orange;
        return this;
    }

    public SnackBuilder alert() {
        param.actionBackgroundResId = SnackBarParam.red;
        return this;
    }

    public SnackBuilder tag(int tag) {
        param.alertTaskId = tag;
        return this;
    }

    public SnackBuilder actionMessageMaxLines(int maxLines) {
        param.actionMessageMaxLine = maxLines;
        return this;
    }

    @SuppressWarnings("Typeface must be defined in assets")
    public SnackBuilder typeface(@NonNull String typeface) {
        param.typeface = typeface;
        return this;
    }

    /**
     * Show.
     */
    public SnackBarParam show() {
        if (snackbar != null)
            snackbar.dismiss();
        if (param.context == null) {
            return param;
        }
        if (param.snackBarView == null) {
            param.snackBarView = param.context.findViewById(android.R.id.content);
        }
        snackbar = Snackbar.make(param.snackBarView, "", param.snackBarDuration);

        // Checked for Message
        if (!TextUtils.isEmpty(param.message)) {
            snackbar.setText(param.message);
        } else {
            snackbar.setText("");
        }

        // checked for ActionMessage
        if (!TextUtils.isEmpty(param.actionMessage)) {
            snackbar.setAction(param.actionMessage, new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (param.onSnackBarActionListener == null) {
                        return;
                    }
                    param.onSnackBarActionListener.onSnackBarActionClicked(param.alertTaskId, v);
                }
            });
        } else {
            snackbar.setAction("", null);
        }
        // checked for actionMessagecolor
        if (param.actionColorResId != -1) {
            snackbar.setActionTextColor(ContextCompat.getColor(param.context, param.actionColorResId));
        }
        TextView tv = (TextView) snackbar.getView().findViewById(android.support.design.R.id.snackbar_text);
        TextView tva = (TextView) snackbar.getView().findViewById(android.support.design.R.id.snackbar_action);
        if (param.actionMessageMaxLine != -1)
            tv.setMaxLines(param.actionMessageMaxLine);
        if (param.textColor != -1) {
            tv.setTextColor(ContextCompat.getColorStateList(param.context, param.textColor));
        }
        if (param.actionBackgroundResId != -1)
            snackbar.getView().setBackgroundColor(ContextCompat.getColor(param.context, param.actionBackgroundResId));
        //check typeface
        String typeface;
        if (!TextUtils.isEmpty(param.typeface)) {
            typeface = param.typeface;
            //set message typeface
            SnackBar.get().setTypeface(param.context, tv, typeface);
            //set action typeface
            SnackBar.get().setTypeface(param.context, tva, typeface);
        }
        snackbar.show();
        return param;
    }

}

File: snackbar/src/main/java/com/snackbar/SnackBuilder.kt
package com.snackbar

import android.app.Activity
import android.support.annotation.ColorRes
import android.support.annotation.StringRes
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.View
import android.widget.TextView

/**
 * Created by clickapps on 2/11/17.
 */

class SnackBuilder constructor(context: Activity) {

    private var param: SnackBarParam = SnackBarParam()
    private var snackbar: Snackbar? = null

    init {
        param.context = context
    }

    constructor(context: Activity, @StringRes resId: Int) : this(context) {
        context?.let {
            param.message = context.getString(resId)
        }
    }

    constructor(context: Activity, msg: String) : this(context) {
        param.message = msg
    }

    constructor(context: Activity, @StringRes resId: Int, @ColorRes backgroundColor: Int) : this(context) {
        param.message = context.getString(resId)
        param.actionBackgroundResId = backgroundColor
    }

    constructor(context: Activity, msg: String, @ColorRes backgroundColor: Int) : this(context) {
        param.message = msg
        param.actionBackgroundResId = backgroundColor
    }

    fun actionMessage(actionMessage: String): SnackBuilder {
        param.actionMessage = actionMessage
        return this
    }

    fun actionMessage(@StringRes resId: Int): SnackBuilder {
        param.context?.let {
            param.actionMessage = param.context!!.getString(resId)
        }
        return this
    }

    fun view(view: View): SnackBuilder {
        param.snackBarView = view
        param.snackBarView?.isFocusable = false
        return this
    }

    fun listener(l: OnSnackBarActionListener): SnackBuilder {
        param.onSnackBarActionListener = l
        return this
    }

    fun actionColor(@ColorRes resId: Int): SnackBuilder {
        param.actionColorResId = resId
        return this
    }

    fun textColor(@ColorRes resId: Int): SnackBuilder {
        param.textColor = resId
        return this
    }

    fun duration(duration: Int): SnackBuilder {
        param.snackBarDuration = duration
        return this
    }

    fun backgroundColor(@ColorRes color: Int): SnackBuilder {
        param.actionBackgroundResId = color
        return this
    }

    fun info(): SnackBuilder {
        param.actionBackgroundResId = SnackBarParam.blue
        return this
    }

    fun confirm(): SnackBuilder {
        param.actionBackgroundResId = SnackBarParam.green
        return this
    }

    fun warning(): SnackBuilder {
        param.actionBackgroundResId = SnackBarParam.orange
        return this
    }

    fun alert(): SnackBuilder {
        param.actionBackgroundResId = SnackBarParam.red
        return this
    }

    fun tag(tag: Int): SnackBuilder {
        param.alertTaskId = tag
        return this
    }

    fun actionMessageMaxLines(maxLines: Int): SnackBuilder {
        param.actionMessageMaxLine = maxLines
        return this
    }

    fun typeface(typeface: String): SnackBuilder {
        param.typeface = typeface
        return this
    }

    /**
     * Show.
     */
    fun show(): SnackBarParam {
        snackbar?.dismiss()
        if (param.context == null) {
            Log.e(javaClass.name, "Context can't be null")
            return param
        }
        var context = param.context!!
        if (param.snackBarView == null) {
            param.snackBarView = context.findViewById(android.R.id.content)
        }
        snackbar = Snackbar.make(param.snackBarView!!, "", param.snackBarDuration)

        // Checked for Message
        if (param.message.isNotEmpty()) {
            snackbar?.setText(param.message)
        } else {
            snackbar?.setText("")
        }

        // checked for ActionMessage
        if (param.actionMessage.isNotEmpty()) {
            snackbar!!.setAction(param.actionMessage, { v ->
                param.onSnackBarActionListener?.onSnackBarActionClicked(param.alertTaskId, v)
            })
        } else {
            snackbar?.setAction("", null)
        }
        // checked for actionMessagecolor
        if (param.actionColorResId != -1) {
            snackbar?.setActionTextColor(ContextCompat.getColor(context, param.actionColorResId))
        }
        val tv = snackbar?.view?.findViewById<View>(android.support.design.R.id.snackbar_text) as TextView
        val tva = snackbar?.view?.findViewById<View>(android.support.design.R.id.snackbar_action) as TextView
        if (param.actionMessageMaxLine != -1)
            tv.maxLines = param.actionMessageMaxLine
        if (param.textColor != -1) {
            tv.setTextColor(ContextCompat.getColorStateList(context, param.textColor))
        }
        if (param.actionBackgroundResId != -1)
            snackbar?.view?.setBackgroundColor(ContextCompat.getColor(context, param.actionBackgroundResId))
        //check typeface
        if (param.typeface.isNotEmpty()) {
            val typeface = param.typeface
            //set message typeface
            SnackBar.get().setTypeface(context, tv, typeface)
            //set action typeface
            SnackBar.get().setTypeface(context, tva, typeface)
        }
        snackbar?.show()
        return param
    }

}
